web/editor_undo_bar.js

```diff
@@ -0,0 +1,101 @@
+/* Copyright 2024 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AnnotationEditorType } from "../src/shared/util.js";
+
+class EditorUndoBar {
+  #container;
+
+  #controller = null;
+
+  isOpen = false;
+
+  #message;
+
+  #undoButton;
+
+  static #l10nMessages = Object.freeze({
+    highlight: "pdfjs-editor-undo-bar-message-highlight",
+    freetext: "pdfjs-editor-undo-bar-message-freetext",
+    stamp: "pdfjs-editor-undo-bar-message-stamp",
+    ink: "pdfjs-editor-undo-bar-message-ink",
+    _multiple: "pdfjs-editor-undo-bar-message-multiple",
+  });
+
+  constructor({ container, message, undoButton, closeButton }, eventBus) {
+    this.#container = container;
+    this.#message = message;
+    this.#undoButton = undoButton;
+
+    // Caveat: we have to pick between registering these everytime the bar is
+    // shown and not having the ability to cleanup using AbortController.
+    const boundHide = this.hide.bind(this);
+    closeButton.addEventListener("click", boundHide);
+    eventBus._on("beforeprint", boundHide);
+    eventBus._on("download", boundHide);
+    eventBus._on("secondarytoolbaraction", boundHide);
+    eventBus._on("annotationeditormodechanged", ({ mode }) => {
+      if (mode === AnnotationEditorType.NONE) {
+        this.hide();
+      }
+    })
```

Comment:
As mentioned in https://github.com/mozilla/pdf.js/pull/18900#pullrequestreview-2379266564, can't this be added directly in the relevant `src/display/editor/`-code so that we don't need an event here!?
The following method seems relevant: https://github.com/mozilla/pdf.js/blob/d37e4b08e4006b382b14b1596c5f70df7e93fbd9/src/display/editor/tools.js#L1616-L1671

Reply:
Similarly, you also want to hide the undo-bar in the following method: https://github.com/mozilla/pdf.js/blob/d37e4b08e4006b382b14b1596c5f70df7e93fbd9/src/display/editor/tools.js#L864-L892